bin/pycbc_compress_bank

```diff
@@ -106,76 +112,66 @@ pycbc.init_logging(args.verbose)
 if args.psd_model or args.psd_file or args.asd_file :
     psd.verify_psd_options(args, parser)
 
+if args.sample_rate % 2 != 0:
+    raise ValueError("sample rate must be a power of 2")
+if args.segment_length % 2 != 0:
+    raise ValueError("segment length must be a power of 2")
+
+df = 1./args.sample_rate
 fmax = args.sample_rate/2.
+N = args.sample_rate * args.segment_length
 
 # load the bank
 logging.info("loading bank")
 # we'll do everything in double precision; if single is desired, we'll
 # cast to single when saving the waveforms
 dtype = numpy.complex128
-# we'll just use dummy values for N, df for now
-bank = waveform.FilterBank(args.bank_file, 5, 0.25, dtype,
+
+bank = waveform.FilterBank(args.bank_file, N/2+1, df, dtype,
                            low_frequency_cutoff=args.low_frequency_cutoff,
                            approximant=args.approximant)
 templates = bank.table
 if args.tmplt_index is not None:
     imin, imax = args.tmplt_index
     templates = templates[imin:imax]
     bank.table = templates
-else:
-    imin, imax = 0, templates.size
-
-# figure out the dfs needed for each waveform
-logging.info("getting needed dfs")
-# we'll ensure that the there are atleast 2 samples in a waveform
-seg_lens = 4*numpy.array([max(4./args.sample_rate,
-    2**numpy.ceil(numpy.log2(compress.rough_time_estimate(m1, m2, flow))))
-    for m1,m2,flow in zip(templates.mass1, templates.mass2, templates.f_lower)])
 
 # generate output file
 logging.info("writing template info to output")
 output = bank.write_to_hdf(args.output, force=args.force)
 
+# get the psd
+logging.info("getting psd")
+psd = pycbc.psd.from_cli(args, length=N/2+1, delta_f=df,
+                         low_frequency_cutoff=templates.f_lower.min(),
+                         dyn_range_factor=pycbc.DYN_RANGE_FAC,
+                         precision='double')
+
 # get the compressed sample points for each template
 logging.info("getting compressed amplitude and phase")
-mismatches = numpy.zeros(imax-imin, dtype=float)
-
-# Create a dictionary to store the psd for every value of N that is used
-psd_dict = {}
-
-for ii in range(imin, imax):
-    # update the N, df to use based on this waveform
-    N = int(args.sample_rate*seg_lens[ii-imin])
-    df = 1./seg_lens[ii-imin]
-    bank.delta_f = df
-    bank.N = N
-    bank.filter_length = N/2 + 1
-    # scratch space
-    decomp_scratch = FrequencySeries(numpy.zeros(N, dtype=dtype), delta_f=df)
+mismatches = numpy.zeros(templates.size, dtype=float)
+
+# scratch space
+decomp_scratch = FrequencySeries(numpy.zeros(N, dtype=dtype), delta_f=df)
```

Comment:
No, the point of the scratch space is to have some allocated memory to decompress the waveform in that you do not need to re-allocate for each waveform. The waveform is decompressed into this memory block, then saved to disk. After that, you're free to re-use the same space for the next template. The reason this was in the for loop before was that `N` was changing from template to template. Now that N is fixed, you can just use the same space for all templates.